source/linux/cpu_usage_sampler.c

```diff
@@ -0,0 +1,140 @@
+
+#include <inttypes.h>
+#include <sys/sysinfo.h>
+#include <sys/types.h>
+
+/*********************************************************************************************************************
+ * Linux Specific
+ ********************************************************************************************************************/
+
+struct aws_cpu_sampler_linux {
+    struct aws_cpu_sampler base;
+
+    uint64_t cpu_last_total_user;
+    uint64_t cpu_last_total_user_low;
+    uint64_t cpu_last_total_system;
+    uint64_t cpu_last_total_idle;
+};
+
+static void s_get_cpu_usage_linux(
+    uint64_t *total_user,
+    uint64_t *total_user_low,
+    uint64_t *total_system,
+    uint64_t *total_idle) {
+
+    FILE *file;
+    int matched_results;
+    file = fopen("/proc/stat", "r");
+    matched_results = fscanf(
+        file,
+        "cpu %" SCNu64 " %" SCNu64 " %" SCNu64 " %" SCNu64 "",
+        (uint64_t *)total_user,
+        (uint64_t *)total_user_low,
+        (uint64_t *)total_system,
+        (uint64_t *)total_idle);
+    fclose(file);
+    if (matched_results == EOF || matched_results != 4) {
+        aws_raise_error(AWS_ERROR_INVALID_STATE);
+    }
+}
+
+static void aws_get_cpu_sample_fn_linux_get_uint64_delta(uint64_t first, uint64_t second, uint64_t *output) {
+    if (first > second) {
+        aws_sub_u64_checked(first, second, output);
+    } else {
+        aws_add_u64_checked((UINT64_MAX - second), first, output);
+    }
+}
+
+static int aws_get_cpu_sample_fn_linux(struct aws_cpu_sampler *sampler, double *output) {
+    struct aws_cpu_sampler_linux *sampler_linux = sampler->impl;
+
+    int return_result = AWS_OP_ERR;
+    uint64_t total_user, total_user_low, total_system, total_idle, total;
+    s_get_cpu_usage_linux(&total_user, &total_user_low, &total_system, &total_idle);
+    // total_combined needs to be double to allow for fractions
+    double percent, total_combined;
+
+    uint64_t total_user_delta = 0, total_user_low_delta = 0, total_system_delta = 0, total_idle_delta = 0;
+    aws_get_cpu_sample_fn_linux_get_uint64_delta(total_user, sampler_linux->cpu_last_total_user, &total_user_delta);
+    aws_get_cpu_sample_fn_linux_get_uint64_delta(
+        total_user_low, sampler_linux->cpu_last_total_user_low, &total_user_low_delta);
+    aws_get_cpu_sample_fn_linux_get_uint64_delta(
+        total_system, sampler_linux->cpu_last_total_system, &total_system_delta);
+    aws_get_cpu_sample_fn_linux_get_uint64_delta(total_idle, sampler_linux->cpu_last_total_idle, &total_idle_delta);
+
+    total_combined = (double)(total_user_delta) + (double)(total_user_low_delta) + (double)(total_system_delta);
+    total = total_combined + (double)(total_idle_delta);
+
+    if (total == 0) {
+        *output = 0;
+        return_result = AWS_OP_ERR;
+        goto cleanup;
+    }
+
+    percent = (total_combined / total) * 100;
+
+    // If negative, there was an error (overflow?)
+    if (percent < 0) {
+        *output = 0;
```

Comment:
Just clamp it to 0. Floating point math is weird and it's actually not hard to go sliiiiiightly below 0 (not that it can happen with the exact math you have here, but in general)

Likewise, be sure to clamp it to the max (I suggested 1.0, but if you stick with 100.0 clamp it to that)